### DISCUSSION TITLE
Netrw-ish File Explorer Today

### DISCUSSION THREAD

**Discussion by baldwindavid** [zed-industries/zed#9067] [Show and tell]
The builtin file drawer in Zed is quite capable and the default vim keymap can make it act somewhat similar to netrw. Hitting `-` will open the drawer at the location of the current file and you can move up and down with `k` and `j` and such.

At the same time, I still find it jarring to look to the left rather than within the current pane I'm focused on. Additionally, when the drawer opens you see the entire hierarchy. Sometimes I can't immediately see where I'm at in the tree.

For those reasons I'm still a big fan of the netrw-ish experience where you open an in-pane directory browser at the current file location and can navigate from there. I used https://github.com/tpope/vim-vinegar in vim and then https://github.com/danprince/vsnetrw in vscode to approximate.

There is no builtin way to do this in Zed but... leveraging the builtin terminal and a custom keybinding can now get us most of the way there.

```json
"-": [
  "workspace::SendKeystrokes",
  ": c r p enter : new center terminal enter ya space cmd-v enter"
],
```

This does the following:
- Opens the command palette
- Copies the current file path (`c r p` ends up as "Copy Relative Path")
- Open the command palette again
- Add a "New Center Terminal" - this will add a terminal in the current pane rather than in the dock at the bottom
- inserts the text "ya " and then paste the current file path in

This opens the [yazi](https://yazi-rs.github.io/) terminal file manager at the location of the current file. You could use any file manager of course (lf, nnn, etc).

While this doesn't carry exactly the same design style as Zed it goes without saying that it is really good at file management. It also has a really good file previewer. I can't quite get image previews to work inside Zed presumably because of Alacritty limitations, but that's okay.

Regardless of the file manager you use, opening a file will create a new tab in the same pane in Zed. That's pretty okay though I'd typically rather have it open in the same tab or at least close the file manager tab. You do end up with some extra tabs. If we eventually have a tab-less Zed option perhaps this problem goes away though.

Anyway, thanks to the Zed team (@osiewicz , @ConradIrwin , @SomeoneToIgnore and more) for continuing to unlock workflows. Just the last couple releases with `SendKeystrokes` and tasks have really opened things up.

**Comment by baldwindavid** [comment]
Here is a recording of usage of the yazi file explorer. It's actually quite a lot better than netrw given the live preview.


https://github.com/zed-industries/zed/assets/4578/ce7baf90-cd67-4404-b824-e9ecaae15759

**Reply by musicq** [reply]
How did you close the terminal tab once open a file from yazi 🤔 ?

**Reply by baldwindavid** [reply]
Good question.

This has evolved a bit and leverages assistance from @sxyazi in a discussion on the yazi repo... https://github.com/sxyazi/yazi/discussions/829

### Keymap

```json
"-": [
  "workspace::SendKeystrokes",
  ": c r p enter cmd-w : new center terminal enter ya_zed space cmd-v enter"
],
```

- `: c r p enter` - open command palette and copy relative path
- `cmd-w` - close current tab
- `: new center terminal enter` - open command palette and add new terminal in current pane
- `ya_zed space cmd-v enter` - run the `ya_zed` command at the commandline with the copied relative path

### `ya_zed`

This is a tweaked version of the shell wrapper from the yazi docs at https://yazi-rs.github.io/docs/quick-start#shell-wrapper provided by @sxyazi

```
function ya_zed() {
	local tmp="$(mktemp -t "yazi-cwd.XXXXX")"
	yazi "$@" --chooser-file="$tmp"

	local opened_file=$(cat -- "$tmp" | head -n 1)
	zed -- "$opened_file"

	rm -f -- "$tmp"
	exit
}
```

This version automatically closes the tab in which yazi is running upon quitting yazi or selecting a file. Thus, it is closing the tab yazi is in, but opening another tab for the selected file.

### Caveat - Clipboard

Since this is reliant upon copying/pasting the relative path to the clipboard, it's going to displace anything on your clipboard. So if you want to copy something, navigate/create another file in yazi, you'll lose that copied text and end up pasting the relative path. A workaround is to set `vim.use_system_clipboard` to "never" so that your vim clipboard isn't impacted. I don't personally love that because I prefer to share the clipboard, but value the vim clipboard staying intact more than keeping in sync with the system clipboard. You could also manage your system clipboard history with 3rd-party software of course.

I think that, ideally, it would be nice to be able to run a task that performs this same behavior targeted at a center terminal similar to what I mentioned in https://github.com/zed-industries/zed/issues/7108#issuecomment-2016542168 - This would allow for grabbing the current path via `$ZED_FILE` and wouldn't leverage copying/pasting at all.

### New Pane

Sometimes I want to keep the current tab intact, but want to open the file explorer at the current location in a new tab. I have a separate keybinding for this with just the slight tweak of not closing the current tab...

```json
"_": [
  "workspace::SendKeystrokes",
  ": c r p enter : new center terminal enter ya_zed space cmd-v enter"
],
```

### Yazi keymap

Lastly, yazi supports keymaps as well. I have bound `-` to go up a level similar to vim-vinegar and `enter` to either enter a directory or open a file.

**Reply by xtrasmal** [reply]
I would like to be able to do that, without enabling vim.
It all works well as a simple task running the ya_zed, but unfortunately it doesn't close the opened terminal window. 
Would be great to allow stacking of commands in keymaps.
- open task
- then after it has run
- close that pane

**Reply by baldwindavid** [reply]
@xtrasmal I don't _think_ any of that is vim-specific. You'd want some sort of chorded keybinding rather than `-` if you are always in insert mode though.

**Reply by xtrasmal** [reply]
I've tried it and couldn't find a way to close the panel after the process in it exits

**Reply by baldwindavid** [reply]
Does it work with vim enabled? If not, there's something else. To be clear, this does not leverage Zed "tasks" or the Zed terminal panel at all and adds that `ya_zed` function to my `.zshrc` and calls it via a keybinding.

**Reply by baldwindavid** [reply]
Everything noted in https://github.com/zed-industries/zed/discussions/9067#discussioncomment-9014433

**Reply by xtrasmal** [reply]
Yes I have followed the ya_zed method. I do the same in Helix using Wezterm's api to open terminal panes(https://quantonganh.com/2023/08/19/turn-helix-into-ide).
Everything works, but the pane does not close. I need another keymap to follow up to activate the pane and close it.

Still have to try it in vim mode. I didn't try vimmode, cause my assumption is that if I would follow your exact steps, I'd get the exact same result. Which I wanted, but without vimmode. 
But to be sure I'll try it just to see the effect.

**Reply by baldwindavid** [reply]
Other thing you might try is to type the same commands as in the keybinding...

```
: c r p enter cmd-w : new center terminal enter ya_zed space cmd-v enter
```

Type each of those things and see that the path is copied, that `cmd-w` closes the tab, that a new center terminal is opened, etc. If it doesn't work by hand it won't work via keybinding either.

**Reply by xtrasmal** [reply]
I got it working. 
I'm not sure what I was thinking, but replacing `:` with `cmd-shift-p` was the obvious move to make it work.

```json
[
  {
    "context": "Workspace",
    "bindings": {
      "cmd-ctrl-alt-shift-z": [
        "workspace::SendKeystrokes",
        "cmd-shift-p crp enter cmd-shift-p wnct enter ya_zed space cmd-v enter"
      ]
    }
  }
]
```

**Reply by baldwindavid** [reply]
Great, glad it's working! This thread will probably help someone in the future too.

**Reply by kakoc** [reply]
I followed the flow and in the end, when I press `enter` over selected element in `yazi` that tab with `yazi` closes and I'm getting back to the place from which I opened `yazi` but the file over which I pressed `enter` in `yazi` wasn't opened. Are there any suggestions which could I missing?

UPD:
I managed to solve the issue by running `cli: install` since `zed` was missing in the terminal originally.

**Reply by jon-walstedt** [reply]
Hi, I am trying this out but it seems that it runs the entire sequence of keystrokes before the terminal has opened so I never get any input in my terminal. Any trick to forwarding the `ya_zed space cmd-v enter` part to the terminal? or is there any delay that i can add if there is a race condition?

**Comment by lionel-** [comment]
> Regardless of the file manager you use, opening a file will create a new tab in the same pane in Zed. That's pretty okay though I'd typically rather have it open in the same tab or at least close the file manager tab. You do end up with some extra tabs. If we eventually have a tab-less Zed option perhaps this problem goes away though.

Maybe a CLI argument for `zed` to open a file in the current tab would get us there? Then we'd set `EDITOR` in the yazi tab to use that argument.

**Comment by baldwindavid** [comment]
fyi, this no longer works as of last week's preview release and looks like no plans for it to work again. See #21610 and #21584

Appears opening in the bottom pane via task is about the closest now.

cc @jon-walstedt

**Comment by baldwindavid** [comment]
Great news. This now can work without any hacky keypress simulation by running a task that opens in a center terminal. This is now in `main` via #22004

**Comment by jon-walstedt** [comment]
Hi @baldwindavid, have you managed to find a  way to open multiple yazi instances  simultaneously?

I have this task:
```
	{
		"label": "ya_zed",
		"command": "ya_zed",
		"use_new_terminal": true,
		"args": ["$ZED_FILE"],
		"shell": "system",
		"hide": "on_success"
	}
```

which i map like this:

```
      "-": [
        "task::Spawn",
        { "task_name": "ya_zed", "reveal_target": "center"}
      ],
```

What i am hoping for is that if i have two (or more) panes and i press "-" it would open yazi in the active pane, and then if i move to the next pane and press "-" again i would want to have yazi open in both panes.

**Reply by baldwindavid** [reply]
@jon-walstedt Yes, just need to set `allow_concurrent_runs`. I use this task...

```json
{
  "label": "File Navigation",
  "command": "ya_zed $ZED_FILE",
  "hide": "always",
  "allow_concurrent_runs": true,
  "use_new_terminal": true
},
```

**Comment by 1925381584** [comment]
I hope that file management should be global, and when you index which file, you can intelligently infer which project you are in. This is exactly what emacs does. If you can do this, you can just open one zed to solve everything.

**Comment by niekdomi** [comment]
Hey,
I have two questions about using Yazi:

1. Is it possible to open Yazi on an empty (untitled) file?  
2. When I open a file like a PDF through Yazi, it closes as soon as I quit Yazi, which I assume is because the PDF's runtime depends on Yazi. However, if I run Yazi in a regular terminal, I can close Yazi, and the PDF remains open. Is there a way to resolve this behavior?

Here’s my `tasks.json` configuration (same as previously discussed):  
```json
{
  "label": "Yazi",
  "command": "yazi_zed $ZED_FILE",
  "hide": "always",
  "use_new_terminal": true,
  "allow_concurrent_runs": false
}
```

And the command being executed:  
```bash
function yazi_zed
    set -l tmp (mktemp -t "yazi-chooser.XXXXX")
    yazi $argv --chooser-file="$tmp"
    set -l opened_file (head -n 1 "$tmp")

    zed -- "$opened_file"
    rm -f -- "$tmp"
end
```

**Comment by tung-dnt** [comment]
this is awesome indeed
